File: apps/zeus-world/engine.js
import { CURATED_2026, SCENARIO_LIBRARY } from "./data/curated-2026.js";

export class CyclicIterator {
  constructor(items) {
    if (!Array.isArray(items) || items.length === 0) {
      throw new Error("CyclicIterator requires at least one item.");
    }
    this.items = items;
    this.index = 0;
  }

  next() {
    const value = this.items[this.index];
    this.index = (this.index + 1) % this.items.length;
    return value;
  }

  take(count) {
    const output = [];
    for (let i = 0; i < count; i += 1) {
      output.push(this.next());
    }
    return output;
  }
}

const ACTIONS = ["cooperate", "compete", "invest", "delay"];
const ACTION_EFFECTS = {
  cooperate: { welfare: 9, risk: -4, cost: 4, fairness: 4 },
  compete: { welfare: 2, risk: 6, cost: 2, fairness: -4 },
  invest: { welfare: 8, risk: -2, cost: 9, fairness: 2 },
  delay: { welfare: -3, risk: 5, cost: 0, fairness: -2 },
};

const ACTOR_PREFERENCES = {
  water: {
    Farmers: { cooperate: 0.7, compete: 0.35, invest: 0.6, delay: -0.2 },
    "City Utility": { cooperate: 0.8, compete: 0.2, invest: 0.5, delay: -0.3 },
    Industry: { cooperate: 0.3, compete: 0.6, invest: 0.45, delay: 0.1 },
    Regulator: { cooperate: 0.9, compete: -0.2, invest: 0.4, delay: 0.2 },
  },
  mobility: {
    Drivers: { cooperate: 0.25, compete: 0.7, invest: 0.3, delay: 0.2 },
    "Transit Agency": { cooperate: 0.85, compete: -0.1, invest: 0.55, delay: 0.1 },
    Businesses: { cooperate: 0.35, compete: 0.55, invest: 0.45, delay: 0.1 },
    Regulator: { cooperate: 0.8, compete: -0.2, invest: 0.6, delay: 0.25 },
  },
  health: {
    Hospitals: { cooperate: 0.8, compete: 0.1, invest: 0.7, delay: -0.4 },
    "Public Agency": { cooperate: 0.9, compete: -0.2, invest: 0.55, delay: 0.1 },
    Residents: { cooperate: 0.45, compete: 0.3, invest: 0.25, delay: 0.25 },
    Regulator: { cooperate: 0.85, compete: -0.25, invest: 0.5, delay: 0.2 },
  },
};

function clamp(value, min, max) {
  return Math.max(min, Math.min(max, value));
}

function deterministicRandom(seed) {
  let value = seed % 2147483647;
  if (value <= 0) {
    value += 2147483646;
  }
  return () => {
    value = (value * 16807) % 2147483647;
    return (value - 1) / 2147483646;
  };
}

function gini(values) {
  if (!values.length) {
    return 0;
  }
  const sorted = [...values].sort((a, b) => a - b);
  const total = sorted.reduce((sum, value) => sum + value, 0);
  if (total <= 0) {
    return 0;
  }
  const weighted = sorted.reduce((sum, value, index) => sum + (index + 1) * value, 0);
  return (2 * weighted) / (sorted.length * total) - (sorted.length + 1) / sorted.length;
}

function selectAction({
  scenario,
  actor,
  pressure,
  incentive,
  penalty,
  collaborationWeight,
  actionCounts,
  rng,
}) {
  const actorPreference = ACTOR_PREFERENCES[scenario]?.[actor] ?? {};
  const actionScores = ACTIONS.map((action) => {
    const effect = ACTION_EFFECTS[action];
    const fatigue = (actionCounts[action] ?? 0) * 0.85;
    const score =
      effect.welfare * (1 + collaborationWeight * 0.2) -
      (effect.risk + pressure * 6) * 0.5 -
      effect.cost * 0.2 +
      (action === "cooperate" || action === "invest" ? incentive * 0.55 : 0) -
      (action === "compete" || action === "delay" ? penalty * 0.45 : 0) +
      (actorPreference[action] ?? 0) * 1.7 -
      fatigue +
      (rng() - 0.5);
    return { action, score };
  });
  actionScores.sort((a, b) => b.score - a.score);
  return actionScores[0].action;
}

function objectiveScore(objectiveFn, { welfare, fairness, risk, cost, budget }) {
  if (objectiveFn === "profit_risk_floor") {
    let score = welfare * 1.1 - risk * 0.75 - Math.max(0, cost - budget) * 1.5;
    if (risk > 45) {
      score -= 30;
    }
    return score;
  }
  if (objectiveFn === "equity_viability_floor") {
    let score = fairness * 1.3 + welfare * 0.45 - risk * 0.6;
    if (welfare < 35) {
      score -= 25;
    }
    return score;
  }
  // welfare_cost_cap
  let score = welfare + fairness * 0.8 - risk * 0.7;
  if (cost > budget) {
    score -= (cost - budget) * 1.8;
  }
  return score;
}

function branchFromLevers(baseLevers) {
  return [
    {
      id: "balanced",
      title: "Balanced Incentives",
      levers: { ...baseLevers },
      pressureDelta: 0,
    },
    {
      id: "incentive_push",
      title: "Incentive-Led Cooperation",
      levers: {
        ...baseLevers,
        budget: baseLevers.budget * 1.08,
        incentive: baseLevers.incentive * 1.25,
        penalty: baseLevers.penalty * 0.9,
        collaborationWeight: baseLevers.collaborationWeight * 1.1,
      },
      pressureDelta: -0.03,
    },
    {
      id: "penalty_push",
      title: "Penalty-Driven Compliance",
      levers: {
        ...baseLevers,
        budget: baseLevers.budget * 0.95,
        incentive: baseLevers.incentive * 0.9,
        penalty: baseLevers.penalty * 1.35,
        collaborationWeight: baseLevers.collaborationWeight * 0.95,
      },
      pressureDelta: 0.03,
    },
    {
      id: "resilience_push",
      title: "Resilience Investment",
      levers: {
        ...baseLevers,
        budget: baseLevers.budget * 1.18,
        incentive: baseLevers.incentive * 1.1,
        penalty: baseLevers.penalty,
        collaborationWeight: baseLevers.collaborationWeight * 1.25,
        rounds: baseLevers.rounds + 1,
      },
      pressureDelta: -0.05,
    },
  ];
}

function simulateBranch({ scenario, baseline, branchConfig, objectiveFn, seed }) {
  const rng = deterministicRandom(seed);
  const stakeholders = SCENARIO_LIBRARY[scenario].stakeholders;
  const rounds = branchConfig.levers.rounds;
  const iterator = new CyclicIterator(stakeholders);
  const payoffs = Object.fromEntries(stakeholders.map((name) => [name, 0]));
  const actionCounts = Object.fromEntries(ACTIONS.map((action) => [action, 0]));
  const branchLog = [];

  let welfareAcc = 0;
  let riskAcc = 0;
  let costAcc = 0;
  let fairnessAcc = 0;

  const pressure = clamp(baseline.pressure + branchConfig.pressureDelta, 0.2, 0.92);

  for (let round = 1; round <= rounds * stakeholders.length; round += 1) {
    const actor = iterator.next();
    const action = selectAction({
      scenario,
      actor,
      pressure,
      incentive: branchConfig.levers.incentive,
      penalty: branchConfig.levers.penalty,
      collaborationWeight: branchConfig.levers.collaborationWeight,
      actionCounts,
      rng,
    });
    actionCounts[action] += 1;

    const effect = ACTION_EFFECTS[action];
    const welfareDelta = effect.welfare * (1 - pressure * 0.25);
    const riskDelta = effect.risk + pressure * 5;
    const costDelta =
      Math.max(0, effect.cost + Math.max(0, branchConfig.levers.incentive - branchConfig.levers.penalty) * 0.18);
    const payoffDelta = welfareDelta - riskDelta * 0.42 - costDelta * 0.22;

    welfareAcc += welfareDelta;
    riskAcc += riskDelta;
    costAcc += costDelta;
    fairnessAcc += effect.fairness;
    payoffs[actor] += payoffDelta;

    branchLog.push(
      `R${String(round).padStart(2, "0")} ${actor} -> ${action} (delta ${payoffDelta.toFixed(2)})`
    );
  }

  const welfare = clamp(welfareAcc / (rounds * 0.65), 0, 100);
  const risk = clamp(riskAcc / (rounds * 1.25), 0, 100);
  const fairness =
    clamp(
      100 - gini(Object.values(payoffs).map((value) => Math.max(0.001, value + 20))) * 100 + fairnessAcc / rounds,
      0,
      100
    );
  const cost = clamp(costAcc, 0, 2000);
  const confidence = clamp(90 - pressure * 22 - Math.abs(branchConfig.levers.incentive - branchConfig.levers.penalty) * 5, 20, 95);
  const objective = objectiveScore(objectiveFn, {
    welfare,
    fairness,
    risk,
    cost,
    budget: branchConfig.levers.budget,
  });

  return {
    id: branchConfig.id,
    title: branchConfig.title,
    welfare: Number(welfare.toFixed(3)),
    fairness: Number(fairness.toFixed(3)),
    risk: Number(risk.toFixed(3)),
    cost: Number(cost.toFixed(3)),
    confidence: Number(confidence.toFixed(3)),
    objectiveScore: Number(objective.toFixed(3)),
    feasible: cost <= branchConfig.levers.budget,
    payoffs: Object.fromEntries(
      Object.entries(payoffs).map(([name, value]) => [name, Number(value.toFixed(3))])
    ),
    branchLog,
  };
}

function runUncertainty({ scenario, baseline, levers, objectiveFn, seed, draws = 40 }) {
  const values = [];
  for (let i = 0; i < draws; i += 1) {
    const drawSeed = seed + i * 17;
    const rng = deterministicRandom(drawSeed);
    const pressureShift = (rng() - 0.5) * 0.16;
    const syntheticBranch = {
      id: `u-${i}`,
      title: "uncertainty",
      levers,
      pressureDelta: pressureShift,
    };
    const draw = simulateBranch({
      scenario,
      baseline,
      branchConfig: syntheticBranch,
      objectiveFn,
      seed: drawSeed,
    });
    values.push(draw.welfare);
  }
  values.sort((a, b) => a - b);
  const idx10 = Math.max(0, Math.floor(values.length * 0.1) - 1);
  const idx90 = Math.min(values.length - 1, Math.ceil(values.length * 0.9) - 1);
  const avg = values.reduce((sum, value) => sum + value, 0) / values.length;
  return {
    meanWelfare: Number(avg.toFixed(3)),
    p10Welfare: values[idx10],
    p90Welfare: values[idx90],
    spread: Number((values[idx90] - values[idx10]).toFixed(3)),
  };
}

function runBacktest({ baseline, recommendedBranch }) {
  const target = baseline.backtest_target;
  const error = Math.abs(recommendedBranch.welfare - target) / Math.max(1, target);
  return {
    target: Number(target.toFixed(3)),
    predicted: recommendedBranch.welfare,
    relativeError: Number(error.toFixed(4)),
    pass: error <= 0.2,
  };
}

function runSensitivity({ scenario, baseline, levers, objectiveFn, seed, baselineScore }) {
  const variants = [
    { incentive: levers.incentive * 0.9, penalty: levers.penalty * 1.1 },
    { incentive: levers.incentive * 1.1, penalty: levers.penalty * 0.9 },
  ];

  const scores = variants.map((variant, index) => {
    const branch = simulateBranch({
      scenario,
      baseline,
      branchConfig: {
        id: `s-${index}`,
        title: "sensitivity",
        levers: { ...levers, ...variant },
        pressureDelta: 0,
      },
      objectiveFn,
      seed: seed + index * 31,
    });
    return branch.objectiveScore;
  });

  const drift = Math.max(...scores.map((value) => Math.abs(value - baselineScore)));
  return {
    objectiveDrift: Number(drift.toFixed(3)),
    pass: drift <= 18,
  };
}

function decisionTree(branches, recommendedId) {
  return branches.map((branch, index) => ({
    node: `N${index + 1}`,
    label: branch.title,
    recommendationPath: branch.id === recommendedId,
    expectedWelfare: branch.welfare,
    expectedCost: branch.cost,
  }));
}

function payoffMatrix(branches, scenario) {
  const stakeholders = SCENARIO_LIBRARY[scenario].stakeholders;
  return stakeholders.map((stakeholder) => {
    const row = { stakeholder };
    branches.forEach((branch) => {
      row[branch.id] = branch.payoffs[stakeholder];
    });
    return row;
  });
}

export async function resolveScenarioData({ scenario, region }) {
  const data = CURATED_2026.regions[region]?.[scenario];
  if (data) {
    return {
      baseline: data,
      sourceChain: ["curated"],
      sourceMode: "curated",
    };
  }

  // Live enrichment placeholder hook.
  try {
    const response = await fetch("https://api.worldbank.org/v2/country/US/indicator/ER.H2O.FWST.ZS?format=json");
    if (response.ok) {
      return {
        baseline: {
          pressure: 0.6,
          backtest_target: 60,
        },
        sourceChain: ["live", "synthetic"],
        sourceMode: "live",
      };
    }
  } catch (error) {
    // Fall back below.
  }

  return {
    baseline: {
      pressure: 0.6,
      backtest_target: 60,
    },
    sourceChain: ["synthetic"],
    sourceMode: "synthetic",
  };
}

export function buildPolicyPack({
  scenario,
  region,
  objectiveFn,
  levers,
  baseline,
  seed = 42,
  peerChecklistComplete = false,
}) {
  const branches = branchFromLevers(levers).map((config, index) =>
    simulateBranch({
      scenario,
      baseline,
      branchConfig: config,
      objectiveFn,
      seed: seed + index * 41,
    })
  );
  const feasible = branches.filter((branch) => branch.feasible);
  const pool = feasible.length > 0 ? feasible : branches;
  const recommended = [...pool].sort((a, b) => b.objectiveScore - a.objectiveScore)[0];

  const uncertainty = runUncertainty({
    scenario,
    baseline,
    levers,
    objectiveFn,
    seed,
  });
  const backtest = runBacktest({ baseline, recommendedBranch: recommended });
  const sensitivity = runSensitivity({
    scenario,
    baseline,
    levers,
    objectiveFn,
    seed,
    baselineScore: recommended.objectiveScore,
  });
  const ready = backtest.pass && sensitivity.pass && peerChecklistComplete && recommended.feasible;

  return {
    runId: `run-${Date.now()}-${Math.floor(seed * 13)}`,
    scenario,
    region,
    objectiveFn,
    branches,
    recommendedBranchId: recommended.id,
    recommendation: recommended,
    uncertainty,
    backtest,
    sensitivity,
    decisionTree: decisionTree(branches, recommended.id),
    payoffMatrix: payoffMatrix(branches, scenario),
    decisionReady: ready,
    negotiationLog: recommended.branchLog,
  };
}

export function scenarioDefaults(scenario) {
  if (scenario === "water") {
    return { budget: 320, incentive: 1.2, penalty: 1.0, collaborationWeight: 1.0, rounds: 8 };
  }
  if (scenario === "mobility") {
    return { budget: 280, incentive: 1.1, penalty: 1.2, collaborationWeight: 1.0, rounds: 8 };
  }
  return { budget: 300, incentive: 1.0, penalty: 1.1, collaborationWeight: 1.05, rounds: 8 };
}

File: AI_Simulations/strategic_policy/src/engine.py
from __future__ import annotations

from dataclasses import asdict, dataclass
import math
import random
from statistics import mean
from typing import Dict, List, Literal, Tuple

from AI_Simulations.common.src.cyclic_iterator import CyclicIterator

ObjectiveFn = Literal[
    "welfare_cost_cap",
    "profit_risk_floor",
    "equity_viability_floor",
]
Domain = Literal["water", "mobility", "health"]

DOMAIN_STAKEHOLDERS: Dict[Domain, List[str]] = {
    "water": ["Farmers", "City Utility", "Industry", "Regulator"],
    "mobility": ["Drivers", "Transit Agency", "Businesses", "Regulator"],
    "health": ["Hospitals", "Public Agency", "Residents", "Regulator"],
}

DOMAIN_BASELINES: Dict[Domain, Dict[str, float]] = {
    "water": {"pressure": 0.66, "backtest_target": 66.0, "risk_floor": 30.0},
    "mobility": {"pressure": 0.59, "backtest_target": 59.0, "risk_floor": 32.0},
    "health": {"pressure": 0.61, "backtest_target": 61.0, "risk_floor": 34.0},
}

ACTIONS = ("cooperate", "compete", "invest", "delay")
ACTION_EFFECTS = {
    "cooperate": {"welfare": 9.0, "cost": 4.0, "risk": -4.0, "equity": 4.0},
    "compete": {"welfare": 2.0, "cost": 2.0, "risk": 6.0, "equity": -4.0},
    "invest": {"welfare": 11.0, "cost": 8.0, "risk": -3.0, "equity": 2.0},
    "delay": {"welfare": -3.0, "cost": 0.0, "risk": 5.0, "equity": -2.0},
}
BACKTEST_MAX_RELATIVE_ERROR = 0.25
SENSITIVITY_MIN_DRIFT = 0.15
SENSITIVITY_MAX_DRIFT = 18.0


@dataclass(frozen=True)
class PolicyLevers:
    budget: float
    incentive: float
    penalty: float
    collaboration_weight: float
    rounds: int = 8


@dataclass(frozen=True)
class BranchResult:
    id: str
    title: str
    welfare: float
    fairness: float
    risk: float
    cost: float
    confidence: float
    objective_score: float
    feasible: bool
    payoffs: Dict[str, float]
    negotiation_log: List[str]


def _clamp(value: float, low: float, high: float) -> float:
    return max(low, min(high, value))


def _gini(values: List[float]) -> float:
    if not values:
        return 0.0
    sorted_values = sorted(values)
    count = len(values)
    total = sum(sorted_values)
    if total <= 0:
        return 0.0
    cumulative = sum((index + 1) * value for index, value in enumerate(sorted_values))
    return (2 * cumulative) / (count * total) - (count + 1) / count


def _action_score(
    action: str,
    pressure: float,
    collaboration_weight: float,
    incentive: float,
    penalty: float,
    rng: random.Random,
) -> float:
    effects = ACTION_EFFECTS[action]
    return (
        effects["welfare"] * (1 + collaboration_weight * 0.2)
        - effects["risk"] * pressure * 0.5
        - effects["cost"] * 0.2
        + (incentive * 0.6 if action in ("cooperate", "invest") else 0.0)
        - (penalty * 0.5 if action in ("compete", "delay") else 0.0)
        + rng.uniform(-0.5, 0.5)
    )


def _choose_best_action(
    pressure: float,
    collaboration_weight: float,
    incentive: float,
    penalty: float,
    rng: random.Random,
) -> str:
    ranked = sorted(
        ACTIONS,
        key=lambda action: _action_score(
            action=action,
            pressure=pressure,
            collaboration_weight=collaboration_weight,
            incentive=incentive,
            penalty=penalty,
            rng=rng,
        ),
        reverse=True,
    )
    return ranked[0]


def _objective_score(
    objective_fn: ObjectiveFn,
    welfare: float,
    fairness: float,
    risk: float,
    cost: float,
    budget: float,
) -> float:
    if objective_fn == "profit_risk_floor":
        score = welfare * 1.1 - risk * 0.8 - max(0.0, cost - budget) * 1.6
        return score if risk <= 45 else score - 30
    if objective_fn == "equity_viability_floor":
        score = fairness * 1.3 + welfare * 0.5 - risk * 0.6
        return score if welfare >= 35 else score - 25
    # welfare_cost_cap
    score = welfare * 1.0 + fairness * 0.8 - risk * 0.7
    if cost > budget:
        score -= (cost - budget) * 1.7
    return score


def _simulate_branch(
    branch_id: str,
    branch_title: str,
    domain: Domain,
    pressure: float,
    objective_fn: ObjectiveFn,
    levers: PolicyLevers,
    seed: int,
) -> BranchResult:
    rng = random.Random(seed)
    stakeholders = DOMAIN_STAKEHOLDERS[domain]
    iterator = CyclicIterator(stakeholders, max_cycles=levers.rounds)
    payoffs = {stakeholder: 0.0 for stakeholder in stakeholders}
    logs: List[str] = []

    aggregate_welfare = 0.0
    aggregate_risk = 0.0
    aggregate_cost = 0.0
    aggregate_equity = 0.0

    for round_index, stakeholder in enumerate(iterator, start=1):
        action = _choose_best_action(
            pressure=pressure,
            collaboration_weight=levers.collaboration_weight,
            incentive=levers.incentive,
            penalty=levers.penalty,
            rng=rng,
        )
        effects = ACTION_EFFECTS[action]
        welfare_delta = effects["welfare"] * (1.0 - pressure * 0.25)
        risk_delta = effects["risk"] + pressure * 5.0
        cost_delta = effects["cost"] + max(0.0, levers.incentive - levers.penalty) * 0.15
        payoff_delta = welfare_delta - risk_delta * 0.45 - cost_delta * 0.2

        aggregate_welfare += welfare_delta
        aggregate_risk += risk_delta
        aggregate_cost += max(0.0, cost_delta)
        aggregate_equity += effects["equity"]
        payoffs[stakeholder] += payoff_delta

        logs.append(
            f"Round {round_index:02d} | {stakeholder} -> {action} | "
            f"payoff_delta={payoff_delta:.2f}"
        )

    # Normalize welfare by an effective-rounds factor so default curated scenarios
    # are calibrated against backtest targets without making weak lever sets pass by default.
    welfare = _clamp(aggregate_welfare / (levers.rounds * 0.65), 0.0, 100.0)
    risk = _clamp(aggregate_risk / levers.rounds, 0.0, 100.0)
    cost = _clamp(aggregate_cost, 0.0, 1000.0)
    equity_component = _clamp(aggregate_equity / levers.rounds, -10.0, 10.0)
    fairness = _clamp(
        100.0 - _gini([max(0.001, value + 20.0) for value in payoffs.values()]) * 100.0 + equity_component,
        0.0,
        100.0,
    )
    confidence = _clamp(
        88.0 - pressure * 20.0 - abs(levers.incentive - levers.penalty) * 4.0,
        25.0,
        95.0,
    )
    objective_score = _objective_score(
        objective_fn=objective_fn,
        welfare=welfare,
        fairness=fairness,
        risk=risk,
        cost=cost,
        budget=levers.budget,
    )
    feasible = cost <= levers.budget

    return BranchResult(
        id=branch_id,
        title=branch_title,
        welfare=round(welfare, 3),
        fairness=round(fairness, 3),
        risk=round(risk, 3),
        cost=round(cost, 3),
        confidence=round(confidence, 3),
        objective_score=round(objective_score, 3),
        feasible=feasible,
        payoffs={key: round(value, 3) for key, value in payoffs.items()},
        negotiation_log=logs,
    )


def _uncertainty_bounds(
    domain: Domain,
    objective_fn: ObjectiveFn,
    levers: PolicyLevers,
    seed: int,
    draws: int = 40,
) -> Dict[str, float]:
    pressure = DOMAIN_BASELINES[domain]["pressure"]
    rng = random.Random(seed + 701)
    welfare_samples: List[float] = []

    for draw in range(draws):
        perturbed_pressure = _clamp(pressure + rng.uniform(-0.08, 0.08), 0.2, 0.9)
        result = _simulate_branch(
            branch_id=f"u-{draw}",
            branch_title="uncertainty-draw",
            domain=domain,
            pressure=perturbed_pressure,
            objective_fn=objective_fn,
            levers=levers,
            seed=seed + draw,
        )
        welfare_samples.append(result.welfare)

    sorted_samples = sorted(welfare_samples)
    low_idx = max(0, int(math.floor(len(sorted_samples) * 0.1)) - 1)
    high_idx = min(len(sorted_samples) - 1, int(math.ceil(len(sorted_samples) * 0.9)) - 1)
    return {
        "mean_welfare": round(mean(welfare_samples), 3),
        "p10_welfare": round(sorted_samples[low_idx], 3),
        "p90_welfare": round(sorted_samples[high_idx], 3),
    }


def _backtest(domain: Domain, predicted: float) -> Dict[str, float | bool]:
    target = DOMAIN_BASELINES[domain]["backtest_target"]
    relative_error = abs(predicted - target) / max(target, 1.0)
    return {
        "target": round(target, 3),
        "predicted": round(predicted, 3),
        "relative_error": round(relative_error, 4),
        "pass": relative_error <= BACKTEST_MAX_RELATIVE_ERROR,
    }


def _sensitivity(
    domain: Domain,
    objective_fn: ObjectiveFn,
    baseline_levers: PolicyLevers,
    seed: int,
) -> Dict[str, float | bool]:
    pressure = DOMAIN_BASELINES[domain]["pressure"]
    base_result = _simulate_branch(
        branch_id="base",
        branch_title="base",
        domain=domain,
        pressure=pressure,
        objective_fn=objective_fn,
        levers=baseline_levers,
        seed=seed,
    )
    perturbed = []
    variants = (
        (0.8, 1.2, 0.9, 0.04, 141),
        (1.25, 0.85, 1.1, -0.04, 212),
    )
    for incentive_mul, penalty_mul, collaboration_mul, pressure_shift, seed_offset in variants:
        altered = PolicyLevers(
            budget=baseline_levers.budget,
            incentive=baseline_levers.incentive * incentive_mul,
            penalty=baseline_levers.penalty * penalty_mul,
            collaboration_weight=baseline_levers.collaboration_weight * collaboration_mul,
            rounds=baseline_levers.rounds,
        )
        result = _simulate_branch(
            branch_id=f"s-{incentive_mul:.2f}",
            branch_title="sensitivity",
            domain=domain,
            pressure=_clamp(pressure + pressure_shift, 0.2, 0.9),
            objective_fn=objective_fn,
            levers=altered,
            seed=seed + seed_offset,
        )
        perturbed.append(result.objective_score)

    drift = max(abs(score - base_result.objective_score) for score in perturbed)
    return {
        "objective_drift": round(drift, 3),
        "pass": SENSITIVITY_MIN_DRIFT <= drift <= SENSITIVITY_MAX_DRIFT,
    }


def build_policy_pack(
    *,
    domain: Domain,
    objective_fn: ObjectiveFn = "welfare_cost_cap",
    budget: float = 320.0,
    incentive: float = 1.2,
    penalty: float = 1.0,
    collaboration_weight: float = 1.0,
    rounds: int = 8,
    seed: int = 42,
    peer_review_checked: bool = False,
) -> Dict[str, object]:
    base_pressure = DOMAIN_BASELINES[domain]["pressure"]
    branch_variants: List[Tuple[str, str, PolicyLevers, float]] = [
        (
            "balanced",
            "Balanced Incentives",
            PolicyLevers(
                budget=budget,
                incentive=incentive,
                penalty=penalty,
                collaboration_weight=collaboration_weight,
                rounds=rounds,
            ),
            base_pressure,
        ),
        (
            "incentive_push",
            "Incentive-Led Cooperation",
            PolicyLevers(
                budget=budget * 1.05,
                incentive=incentive * 1.25,
                penalty=penalty * 0.9,
                collaboration_weight=collaboration_weight * 1.1,
                rounds=rounds,
            ),
            _clamp(base_pressure - 0.03, 0.2, 0.9),
        ),
        (
            "penalty_push",
            "Compliance Penalty Focus",
            PolicyLevers(
                budget=budget * 0.95,
                incentive=incentive * 0.85,
                penalty=penalty * 1.35,
                collaboration_weight=collaboration_weight * 0.95,
                rounds=rounds,
            ),
            _clamp(base_pressure + 0.03, 0.2, 0.9),
        ),
        (
            "resilience",
            "Resilience Investment",
            PolicyLevers(
                budget=budget * 1.15,
                incentive=incentive * 1.1,
                penalty=penalty,
                collaboration_weight=collaboration_weight * 1.25,
                rounds=rounds + 1,
            ),
            _clamp(base_pressure - 0.05, 0.2, 0.9),
        ),
    ]

    branches: List[BranchResult] = []
    for index, (branch_id, title, branch_levers, pressure) in enumerate(branch_variants):
        branches.append(
            _simulate_branch(
                branch_id=branch_id,
                branch_title=title,
                domain=domain,
                pressure=pressure,
                objective_fn=objective_fn,
                levers=branch_levers,
                seed=seed + index * 37,
            )
        )

    feasible = [branch for branch in branches if branch.feasible]
    recommendation_pool = feasible if feasible else branches
    recommended = max(recommendation_pool, key=lambda branch: branch.objective_score)

    payoff_matrix = {
        stakeholder: {branch.id: branch.payoffs.get(stakeholder, 0.0) for branch in branches}
        for stakeholder in DOMAIN_STAKEHOLDERS[domain]
    }

    uncertainty = _uncertainty_bounds(
        domain=domain,
        objective_fn=objective_fn,
        levers=PolicyLevers(
            budget=budget,
            incentive=incentive,
            penalty=penalty,
            collaboration_weight=collaboration_weight,
            rounds=rounds,
        ),
        seed=seed,
    )
    backtest = _backtest(domain=domain, predicted=recommended.welfare)
    sensitivity = _sensitivity(
        domain=domain,
        objective_fn=objective_fn,
        baseline_levers=PolicyLevers(
            budget=budget,
            incentive=incentive,
            penalty=penalty,
            collaboration_weight=collaboration_weight,
            rounds=rounds,
        ),
        seed=seed,
    )

    decision_ready = bool(
        backtest["pass"]
        and sensitivity["pass"]
        and peer_review_checked
        and recommended.feasible
    )

    return {
        "domain": domain,
        "objective_fn": objective_fn,
        "branches": [asdict(branch) for branch in branches],
        "recommended_branch_id": recommended.id,
        "payoff_matrix": payoff_matrix,
        "uncertainty": uncertainty,
        "backtest": backtest,
        "sensitivity": sensitivity,
        "decision_ready": decision_ready,
    }
